app/src/main/scala/org/alephium/explorer/service/TransactionService.scala

```diff
@@ -221,71 +217,9 @@ object TransactionService extends TransactionService {
       intervalType: IntervalType,
       paralellism: Int
   )(implicit ec: ExecutionContext, dc: DatabaseConfig[PostgresProfile]): Flowable[Buffer] = {
-    val timeranges = amountHistoryTimeRanges(from, to, intervalType)
-    val amountHistory = Flowable
-      .fromIterable(timeranges.asJava)
-      .concatMapEager(
-        { case (from: TimeStamp, to: TimeStamp) =>
-          Flowable.fromCompletionStage(getInOutAmount(address, from, to).asJava)
-        },
-        paralellism,
-        1
-      )
-      .filter {
-        // Ignore time ranges without data
-        case (in, out, _) => in != U256.Zero || out != U256.Zero
-      }
-      .scan(
-        (BigInteger.ZERO, TimeStamp.zero),
-        { (acc: (BigInteger, TimeStamp), next) =>
-          val (sum, _)      = acc
-          val (in, out, to) = next
-          val diff          = out.v.subtract(in.v)
-          val newSum        = sum.add(diff)
-          (newSum, to)
-        }
-      )
-      .skip(1) // Drop first elem which is the seed of the scan (0,0)
-
-    amountHistoryToJsonFlowable(
-      amountHistory
-    )
-  }
-
-  def amountHistoryToJsonFlowable(history: Flowable[(BigInteger, TimeStamp)]): Flowable[Buffer] = {
-    history
-      .concatMap { case (diff, to: TimeStamp) =>
-        val str = s"""[${to.millis},"$diff"]"""
-        Flowable.just(str, ",")
-      }
-      .skipLast(1) // Removing latest "," it replace the missing `intersperse`
-      .startWith(Flowable.just("""{"amountHistory":["""))
-      .concatWith(Flowable.just("]}"))
-      .map(Buffer.buffer)
-  }
-
-  @SuppressWarnings(Array("org.wartremover.warts.OptionPartial"))
-  private def amountHistoryTimeRanges(
-      from: TimeStamp,
-      to: TimeStamp,
-      intervalType: IntervalType
-  ): ArraySeq[(TimeStamp, TimeStamp)] = {
-    val fromTruncated =
-      TimeStamp.unsafe(
-        Instant
-          .ofEpochMilli(
-            if (from.isBefore(ALPH.LaunchTimestamp)) ALPH.LaunchTimestamp.millis else from.millis
-          )
-          .truncatedTo(intervalType.chronoUnit)
-          .toEpochMilli
-      )
-
-    val first = (ALPH.GenesisTimestamp, fromTruncated.minusUnsafe(Duration.ofMillisUnsafe(1)))
-    first +: TimeUtil.buildTimestampRange(
-      fromTruncated,
-      to,
-      (intervalType.duration - Duration.ofMillisUnsafe(1)).get
-    )
+    FlowableUtil.getAmountHistory(from, to, intervalType, paralellism) { case (fromTs, toTs) =>
+      getInOutAmount(address, fromTs, toTs)
```

Comment:
Another question: Since this query `getInOutAmount` is being invoked multiple times for a single `address` with different timestamps ranges, would it simpler and more efficient to fetch this data in a single query? Since `address` and `timestamps` are both indexed.

For example (untested):
```sql
SELECT SUM(output_ref_amount)
      FROM inputs
WHERE ...
    AND (block_timestamp BETWEEN $fromTime1 AND $toTime1)
    OR (block_timestamp BETWEEN $fromTime2 AND $toTime2)
...
```
Totally understandable if reactive stream is preferred instead, for back-pressure and other fancy stuff that comes with it.

Reply:
very good point, as this query is limited to max 365 days, I could give a try and check performance. Backpresure and streaming is more important for the txs exporting endpoint.